### PR TITLE
Solve scan build bugs

### PR DESCRIPTION
**Pull request by evintila**
Solved the scan-build detected bugs that were not considered false positives -- mainly caused by scan-build ignoring THROWEXCEPTION.

### CODE REVIEW

File: src/core/Cfg.cpp [hunk 1]
```diff
@@ -25,7 +25,12 @@ std::string CfgBase::_get(const std::string& name, XMLElement* elem)
 
 std::string CfgBase::get(const std::string& name, XMLElement* elem)
 {
-	if (!elem) elem = _elem;
+	if (!elem)
+	{
+	    if (!_elem)
+	        THROWEXCEPTION("Error: accessing null XML element");
+		elem = _elem;
```
Comment (nickbroon) [inline src/core/Cfg.cpp]: The indentation needs fixed here to make it clear that `elem = _elem` is happen irrespective of the if statement. Or use an `else`

**Comment by nickbroon** [comment]
17f9626 should probably be squashed into 85887c8

File: src/core/Cfg.cpp [hunk 2]
```diff
@@ -25,7 +25,13 @@ std::string CfgBase::_get(const std::string& name, XMLElement* elem)
 
 std::string CfgBase::get(const std::string& name, XMLElement* elem)
 {
-	if (!elem) elem = _elem;
+	if (!elem)
+	{
+		if (!_elem)
+			THROWEXCEPTION("Error: accessing null XML element");
+		else
```
Comment (ogasser) [inline src/core/Cfg.cpp]: Please use braces around if and else blocks for all new code.

File: src/modules/analysis/fpsigmatcher/Matcher.c [hunk 3]
```diff
@@ -73,7 +73,16 @@ struct Mapper * new_mapper(struct Matcher * matcher){
 
 	mapper->complete_numOfTokens = complete_numOfTokens;
 	mapper->classwise_numOfTokens = malloc(sizeof(int)*matcher->numOfClasses);
-	mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	if (complete_numOfTokens != 0)
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Matcher.c]: Move brace to if line

File: src/modules/analysis/fpsigmatcher/Matcher.c [hunk 4]
```diff
@@ -73,7 +73,16 @@ struct Mapper * new_mapper(struct Matcher * matcher){
...
+	    mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	}
+	else
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Matcher.c]: Move brace to else line

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 5]
```diff
@@ -27,8 +27,15 @@ struct BayesSignature * new_Signature_i(int maxNumOfTokens, char * filename){
 	struct BayesSignature * signature;
 	signature = (struct BayesSignature *)malloc(sizeof(struct BayesSignature));
 	// Allocate memory for the pointers to the actual Tokens
-	signature->tokens = (struct Token **)malloc(maxNumOfTokens*sizeof(struct Token *));
-
+	if (maxNumOfTokens != 0)
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move brace up

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 6]
```diff
@@ -27,8 +27,15 @@ struct BayesSignature * new_Signature_i(int maxNumOfTokens, char * filename){
...
+		signature->tokens = (struct Token **)malloc(maxNumOfTokens*sizeof(struct Token *));
+	}
+	else
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move brace up

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 7]
```diff
@@ -79,8 +86,16 @@ struct BayesSignature * new_Signature_s(const char * absolute_path, char * filen
 
 	/** Allocate Memory for the **file and for the tokens and supports */
 	int i = 0;
-	char ** lines = (char **)malloc(numOfLines * sizeof(char *));
-
+	char ** lines;
+	if (numOfLines != 0)
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move up

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 8]
```diff
@@ -79,8 +86,16 @@ struct BayesSignature * new_Signature_s(const char * absolute_path, char * filen
...
+		lines = (char **)malloc(numOfLines * sizeof(char *));
+	}
+	else
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move up

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 9]
```diff
@@ -79,8 +86,16 @@ struct BayesSignature * new_Signature_s(const char * absolute_path, char * filen
...
 	/** Allocate Memory for the **file and for the tokens and supports */
 	int i = 0;
-	char ** lines = (char **)malloc(numOfLines * sizeof(char *));
-
+	char ** lines;
+	if (numOfLines != 0)
+	{
+		lines = (char **)malloc(numOfLines * sizeof(char *));
+	}
+	else
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move up

File: src/modules/analysis/fpsigmatcher/Matcher.c [hunk 10]
```diff
@@ -73,7 +73,16 @@ struct Mapper * new_mapper(struct Matcher * matcher){
 
 	mapper->complete_numOfTokens = complete_numOfTokens;
 	mapper->classwise_numOfTokens = malloc(sizeof(int)*matcher->numOfClasses);
-	mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	if (complete_numOfTokens != 0)
+	{
+	    mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	}
+	else
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Matcher.c]: Move else to closing brace

File: src/modules/analysis/fpsigmatcher/Signature.c [hunk 11]
```diff
@@ -27,8 +27,15 @@ struct BayesSignature * new_Signature_i(int maxNumOfTokens, char * filename){
 	struct BayesSignature * signature;
 	signature = (struct BayesSignature *)malloc(sizeof(struct BayesSignature));
 	// Allocate memory for the pointers to the actual Tokens
-	signature->tokens = (struct Token **)malloc(maxNumOfTokens*sizeof(struct Token *));
-
+	if (maxNumOfTokens != 0)
+	{
+		signature->tokens = (struct Token **)malloc(maxNumOfTokens*sizeof(struct Token *));
+	}
+	else
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Signature.c]: Move else up

File: src/modules/analysis/fpsigmatcher/Utils.c [hunk 12]
```diff
@@ -41,32 +41,43 @@ double get_double(char string[]){
...
 		char 	separator[] = ",. ";
-		strtoken(string, separator, tokens);
-		//printf("STRINGS: %s  ,  %s\n", tokens[0], tokens[1]);
+		int num_tokens = strtoken(string, separator, tokens);
 
-		//sscanf(string, "%d %d", &pre_comma, &post_comma);
+		if (num_tokens < 1){
+			// no tokens -> cannot compose a number
+		}
+		else if (num_tokens == 1){
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Utils.c]: Move up

File: src/modules/analysis/fpsigmatcher/Utils.c [hunk 13]
```diff
@@ -41,32 +41,43 @@ double get_double(char string[]){
...
+			// only one token
+			ret = atoi(tokens[0]);
+		}
+		else{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Utils.c]: Move up

File: src/modules/analysis/fpsigmatcher/Utils.c [hunk 14]
```diff
@@ -127,7 +138,16 @@ char ** get_filenames(const char * directory, int * num_of_files) {
 	/* Get each directory entry from pDIR and store its name
 	 * in an array */
 	char ** filenames;
-	filenames = (char **)malloc(sizeof(char*)*filecounter);
+	if ( filecounter != 0)
+	{
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Utils.c]: Move up

File: src/modules/analysis/fpsigmatcher/Utils.c [hunk 15]
```diff
@@ -127,7 +138,16 @@ char ** get_filenames(const char * directory, int * num_of_files) {
...
+	    filenames = (char **)malloc(sizeof(char*)*filecounter);
+	}
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Utils.c]: Combine with else and { below

File: src/modules/ipfix/IpfixRecord.cpp [hunk 16]
```diff
@@ -35,7 +35,7 @@ namespace InformationElement {
 		if (length==0) {
 			const ipfix_identifier* ipfixid = ipfix_id_lookup(id, enterprise);
 			if (ipfixid)
```
Comment (ogasser) [inline src/modules/ipfix/IpfixRecord.cpp]: Add braces to if

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 17]
```diff
@@ -94,14 +94,15 @@ Rule* AggregatorBaseCfg::readRule(XMLElement* elem) {
...
-				msg(MSG_ERROR, "AggregatorBaseCfg: Match pattern for id=%d ignored because biflow aggregation is enabled.", rule->field[i]->type.id);
-				free(rule->field[i]->pattern);
-				rule->field[i]->pattern = NULL;
+	else {
+		// exclude coexistence of patterns and biflow aggregation
+		if(rule->biflowAggregation) {
+			for(int i=0; i < rule->fieldCount; i++) {
+				if(rule->field[i]->pattern)
+					msg(MSG_ERROR, "AggregatorBaseCfg: Match pattern for id=%d ignored because biflow aggregation is enabled.", rule->field[i]->type.id);
+					free(rule->field[i]->pattern);
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: Indentation does not match -> Add braces to make clear what is part of the if body.

File: src/modules/packet/filter/RandomSampler.cpp [hunk 18]
```diff
@@ -39,13 +39,9 @@ RandomSampler::RandomSampler(int n, int N) : samplingSize(N), acceptSize(n), cur
         int pos;
 
         if(n > N) {
-                int tmp;
-
                 msg(MSG_ERROR, "%d out-of %d makes no sense - exchanging values", n, N);
 
-                tmp=n;
```
Comment (ogasser) [inline src/modules/packet/filter/RandomSampler.cpp]: Shouldn't we update samplingSize and acceptSize as well?

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp (hunk 17 above)
Comment (evintila) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: I think the original code was wrong here, and the free should only be in the then branch

Comment (ogasser) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: As I interpret the comment, it says that biflow aggregation in combination with pattern matching is not supported. Therefore, if we find a pattern with biflow aggreagation, we throw the error, free() and set to NULL. What do you think @evintila?

Comment (evintila) [reply src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: I agree

File: src/modules/analysis/fpsigmatcher/Matcher.c [hunk 19]
```diff
@@ -73,7 +73,14 @@ struct Mapper * new_mapper(struct Matcher * matcher){
 
 	mapper->complete_numOfTokens = complete_numOfTokens;
 	mapper->classwise_numOfTokens = malloc(sizeof(int)*matcher->numOfClasses);
-	mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	if (complete_numOfTokens != 0) {
+	    mapper->supports = malloc(sizeof(double)*complete_numOfTokens);
+	}
+	else {
```
Comment (ogasser) [inline src/modules/analysis/fpsigmatcher/Matcher.c]: move else up

File: src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp [hunk 20]
```diff
@@ -94,14 +94,16 @@ Rule* AggregatorBaseCfg::readRule(XMLElement* elem) {
...
 	}
-
-	// exclude coexistence of patterns and biflow aggregation
-	if(rule->biflowAggregation) {
-		for(int i=0; i < rule->fieldCount; i++) {
-			if(rule->field[i]->pattern)
-				msg(MSG_ERROR, "AggregatorBaseCfg: Match pattern for id=%d ignored because biflow aggregation is enabled.", rule->field[i]->type.id);
-				free(rule->field[i]->pattern);
-				rule->field[i]->pattern = NULL;
+	else {
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/AggregatorBaseCfg.cpp]: move else up

File: src/modules/ipfix/aggregator/FlowHashtable.cpp [hunk 21]
```diff
@@ -738,10 +735,9 @@ void FlowHashtable::aggregateDataRecord(IpfixDataRecord* record)
 				default:
 					break;
 			}
-			continue;
 		}
-
-		if (!fieldFilled) {
+		else {
```
Comment (ogasser) [inline src/modules/ipfix/aggregator/FlowHashtable.cpp]: move else up

File: src/modules/packet/PCAPExporterPipe.cpp [hunk 22]
```diff
@@ -109,6 +109,11 @@ int PCAPExporterPipe::execCmd(std::string& cmd)
 	}
 	strcpy(tmp, cmd.c_str());
 	command[0] = strtok(tmp, " \t");
+	if (NULL == command[0])
+	{
```
Comment (ogasser) [inline src/modules/packet/PCAPExporterPipe.cpp]: move brace up

File: src/modules/packet/filter/RandomSampler.cpp (hunk 18 above)
Comment (ogasser) [reply src/modules/packet/filter/RandomSampler.cpp]: Done ✅

**Pull request merged**